src/runtime_filesystem.cpp

```
@@ -162,42 +162,81 @@ int64_t __ry_filesystem_copy(const char *src, const char *dst) {
-    if (stat(src, &src_st) != 0) {
+    if (fstat(src_fd, &src_st) != 0) {
         setLastError("copy: cannot stat source '%s': %s", src, strerror(errno));
+        close(src_fd);
         return 1;
     }
-    FILE *in = fopen(src, "rb");
-    if (!in) {
-        setLastError("copy: cannot open source '%s': %s", src, strerror(errno));
+    // Open without O_TRUNC first, then check for same-inode to prevent
+    // zeroing the source when src and dst refer to the same file.
+    int dst_fd = open(dst, O_WRONLY | O_CREAT, src_st.st_mode & 07777);
```

Clang (14.0.6)

[warning] variable 'dst_fd' is not initialized

(cppcoreguidelines-init-variables)